### ISSUE TITLE
`Can't find ref com.submlimetext.three` error

### ISSUE THREAD

**Issue by trappedinspacetime** [rmacqueen/sublime-flatpak#7]
First of all thank you for your work. I wanted to install `flatpak sublime` package, I cloned this repo followed the steps in `readme` page. 
I ran `make` then added repository `flatpak --user remote-add --no-gpg-verify local-sublime repo` 
but 
`flatpak --user install local-sublime com.submlimetext.three` gives the error below:

      Error searching remote local-sublime: Can't find ref com.submlimetext.three

I am new to `flatpak`, what am I doing wrong?

Regards
Kenn

**Comment by ptomato** [comment]
What is the output of `flatpak remotes -d`? It should show the path for the `local-sublime` repository that you added. If that seems OK, then what is the output of `flatpak remote-ls local-sublime`? That should show all the apps that are available for install from the `local-sublime` repository.

**Comment by trappedinspacetime** [comment]
@ptomato thank you for responding. 

      flatpak remotes -d | grep sublime
prints

      local-sublime     -     file:///home/***/CODING-RELOADED/FLATPAK/sublime-flatpak/repo 1        user,no-gpg-verify

and 

      flatpak remote-ls local-sublime
prints

      Remote "local-sublime" not found

By the way, I am on Ubuntu 16.04 32 bit installation but running it with 64 bit kernel.

**Comment by ptomato** [comment]
Well, that's odd - it's listing `local-sublime` as a remote (I assume you checked that `/home/***/CODING-RELOADED/FLATPAK/sublime-flatpak/repo` exists?) but then says it doesn't exist. Try removing and re-adding the remote (look at the help for `flatpak remote-delete` and `flatpak remote-add`.)

**Comment by trappedinspacetime** [comment]
I deleted the repo with 

        flatpak --user remote-delete local-sublime
then I removed cloned `sublime-flatpak` directory. I followed the steps from clean cloned `sublime-flatpak` .
Unfortunately the result is the same.
By the way I set terminal environment to i386 with `setarch i386 bash` to prevent architecture conflicts.

**Comment by mwleeds** [comment]
@kensully you need to run `flatpak --user remote-ls -d local-sublime`. Should be fixed soon (https://github.com/flatpak/flatpak/pull/1205)

**Comment by trappedinspacetime** [comment]
@mwleeds  thank you for the info. It lets me add `sublime-flatpak` repository and I installed the application by ` flatpak --user install local-sublime com.sublimetext.three` then I run application with `flatpak run com.sublimetext.three` but I get the error following:

      /bin/sh: warning: setlocale: LC_ALL: cannot change locale (C.UTF-8)
     cp: cannot create regular file '/home/****/.var/app/com.sublimetext.three/config/sublime-text-3/Installed Packages': No such file or directory
     /bin/sh: warning: setlocale: LC_ALL: cannot change locale (C.UTF-8)
     /app/extra/bin/subl: line 2: /app/extra/opt/sublime_text/sublime_text: No such file or directory

**Comment by ptomato** [comment]
Looks like it's failing to run the "extra data" script that it's supposed to do on first run.

Can you check if each component of this path exists? `/home/****/.var/app/com.sublimetext.three/config/sublime-text-3/Installed Packages`
Also, check if `/home/****/.local/share/flatpak/app/com.sublimetext.three/current/active/files/extra/opt/sublime_text/sublime_text` exists.

**Comment by trappedinspacetime** [comment]
Do I need to do `make install` after `make`?
`/home/****/.var/app/com.sublimetext.three/config/sublime-text-3/` does not exist but
 `/home/****/.local/share/flatpak/app/com.sublimetext.three/current/active/files/extra/opt/sublime_text/sublime_text` exists.

**Comment by ptomato** [comment]
`make install` should not be necessary. Try creating that directory: `~/.var/app/com.sublimetext.three/config/sublime-text-3/Installed\ Packages`. That might allow it to succeed. I'm not sure what exactly is failing for lack of that directory, though, so I don't know how to fix it more permanently.

**Comment by trappedinspacetime** [comment]
I created that directory but it still gives me the same error:

      /app/extra/bin/subl: line 2: /app/extra/opt/sublime_text/sublime_text: No such file or directory

**Comment by ptomato** [comment]
Could you check the output of `flatpak --version`? Your version of flatpak might be too old to support downloading and installing the .deb file during install.

**Comment by trappedinspacetime** [comment]
It's `Flatpak 0.9.98`

**Comment by ptomato** [comment]
Sorry that you haven't heard anything in a while, but at this point I have no idea what the problem could be... I'll let you know if I can think of anything.

**Comment by trappedinspacetime** [comment]
@ptomato thank you for your heartiness on being helpful to me. I'll wait for ` Flatpak 0.10.2` release then I ll try again.

**Comment by ramcq** [comment]
> By the way, I am on Ubuntu 16.04 32 bit installation but running it with 64 bit kernel.
> By the way I set terminal environment to i386 with setarch i386 bash to prevent architecture conflicts.

` /app/extra/bin/subl: line 2: /app/extra/opt/sublime_text/sublime_text: No such file or directory`

I think this error is a kernel exec format error meaning that it failed to find the dynamic linker for this binary, which is probably some kind of architecture mismatch caused by your setup.

It's a 64-bit binary, so we need Flatpak to build and run it in a 64-bit runtime - it should be entirely possible for Flatpak to do this 64-bit stuff inside a container given your 64-bit kernel, but it's possible your setarch has messed it up and made it select a 32-bit runtime.

Ideas:
- Check the output of `flatpak --supported-arches` - if you don't see x86_64 there, this won't work.
- Try it without the setarch and see what flatpak-builder and flatpak does by itself.
- Pass --arch=x86_64 to flatpak-builder and flatpak when handling this package.
- Reinstall with 64-bit and allow multiarch and/or containers to handle any 32-bit apps you have - this configuration is far more likely to have been tested.

@ptomato I wonder if there's a way to mark the whole app as x86_64? I see modules and build-options can be specified by arch, but no way to make Flatpak just fail if you try and build on non x86_64.

**Comment by trappedinspacetime** [comment]
@ramcq I am grateful to your thorough insight and guidence.
 
      flatpak --supported-arches
      i386

 I  switched to `32 bit kernel again` to test your suggestions. I tested it in 32 bit environment 

       flatpak run com.sublimetext.three/i386/1.0
       /bin/sh: warning: setlocale: LC_ALL: cannot change locale (C.UTF-8)
       /bin/sh: warning: setlocale: LC_ALL: cannot change locale (C.UTF-8)
       /app/extra/bin/subl: line 2: /app/extra/opt/sublime_text/sublime_text: cannot execute binary file: Exec format error
       /app/extra/bin/subl: line 2: /app/extra/opt/sublime_text/sublime_text: Success  
then
 I deleted `local-sublime` from `flatpak` and rebuilt the `sublime-flatpak` but in the final process I got the following error:

       min-free-space-percent '3%' would be exceeded, 23.4 GB available.
I removed some bulk files and ran `make` again (I don't know if I should have removed  `sublime-flatpak` and built it from scratch again after that space error) , `make` finalized without error. 

But now I get this warning:

       flatpak --user remote-add --no-gpg-verify local-sublime repo
       Warning: Could not update extra metadata for 'local-sublime': Remote listing for local-sublime not available; server has no summary file
       Check the URL passed to remote-add was valid

and I get the exact same error which I reported the issue:

      flatpak --user install local-sublime com.submlimetext.three
      Error searching remote local-sublime: Can't find ref com.submlimetext.three

**Comment by trappedinspacetime** [comment]
I installed it following way

       flatpak --user remote-ls -d local-sublime
       Ref                                Commit       Installed size Download size
       app/com.sublimetext.three/i386/1.0 9f3642061cda 37.9 MB        22.1 MB
       flatpak --user install local-sublime com.sublimetext.three/i386/1.0

and I get the error:

       flatpak run com.sublimetext.three
       /bin/sh: warning: setlocale: LC_ALL: cannot change locale (C.UTF-8)
       /bin/sh: warning: setlocale: LC_ALL: cannot change locale (C.UTF-8)
       /app/extra/bin/subl: line 2: /app/extra/opt/sublime_text/sublime_text: cannot execute binary file: Exec format error
       /app/extra/bin/subl: line 2: /app/extra/opt/sublime_text/sublime_text: Success

I also tried this with no luck:

       flatpak run --runtime=org.gnome.Sdk/i386/3.24 com.sublimetext.three

**Comment by trappedinspacetime** [comment]
I've found the culprit. It only provides links for `sublimetext` 64-bit.

       com.sublimetext.three.json
       15:        "--extra-data=sublime.deb:ecd78f6fd3a61dc2d68725a368343589d8df4193434c0275cd49b2026fd0fb81:8190414::https://download.sublimetext.com/files/sublime-text_build-3143_amd64.deb",

        sublime/files/manifest.json
        91:    "--extra-data=sublime.deb:ecd78f6fd3a61dc2d68725a368343589d8df4193434c0275cd49b2026fd0fb81:8190414::https://download.sublimetext.com/files/sublime-text_build-3143_amd64.deb",

I changed it to 

       https://download.sublimetext.com/files/sublime-text_build-3143_i386.deb

but package  control fails.

**Comment by nedrichards** [comment]
Right! there's a new extra-data syntax for the manifest that makes it easier to offer different files for different architectures. Would probably be worth switching to that @ptomato after Endless 3.3.5 is released which comes with the new flatpak and flatpak-builder.

**Comment by ramcq** [comment]
I'm sure I said earlier that this Flatpak was a 64-bit binary and given you have a 64-bit kernel, you can simply make Flatpak run that. As I said, if you have a 64-bit kernel, if you get rid of your setarch stuff and allow flatpak to run a 64-bit sandbox, or pass --arch=x86_64, you should be able run this 64-bit flatpak unmodified, even though your system outside is 32-bit userland (Flatpak doesn't care).

**Comment by trappedinspacetime** [comment]
@ramcq  At first I thought `sublimetext` provided only `32-bit` binary, that's why I struggled to install it into`32-bit` environment. 

> It's a 64-bit binary, so we need Flatpak to build and run it in a 64-bit runtime 

I think I missed that line or I misinterpreted it.

Anyway I fixed it for `32-bit sublimetext` binary by replacing corresponding lines with 

      sublime.deb:116a237e634ee0b0a659295c141567ceb99c33d1b256905a680a192cf08c229f:8138198::https://download.sublimetext.com/files/sublime-text_build-3143_i386.deb

in `com.sublimetext.three.json` and `sublime/files/manifest.json`

I just reran `make` and added repo then finally installed package.
It works well.

Thank you for all your help.
Regards.
